162 add reordering to search columns

tedschu opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Added the ability to sort (desc, asc) on numeric data columns in the search results. Specifically:

**In Search.vue component:**
- Updated with new state values to track the field being sorted (ex. 'GHGEmissions'), direction of sort (ex. 'asc'), and to track when there are filtered results, so that the sort can only apply to those (ex. if 'propertyTypeFilter' is applied).
- Added handleSort function to accept param 'field' (e.g. column that has been clicked) and update sortedField and sortedDirection state values, and to pass either all buildings or filtered buildings to be filtered.
- Added runSort function, which takes a buildings param from handleSort and sorts by the selected field, ascending or descending.

**In BuildingsTable.vue component:**
- Added props for sortedField and sortedDirection from Search component
- Added click emitters for all numeric data columns (including those that show conditionally, such as 'showGasUse').
- Added sort indicator to show asc || desc, with basic styling (ex. added cursor on hover).

Descending sort on GHG Emissions column:
![image](https://github.com/user-attachments/assets/2b8c7539-7208-435c-ada7-336c0efe0395)
 
Fixes #162 

# Testing Instructions

Please describe the tests/QA that you did to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration

- I tested on all user states (ex. no search value, filters applied, etc.) and the sorting applied correctly. 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

<!-- PR template modified from: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ -->

vkoves commented:
@tedschu - this is awesome! The one thing is it's not clear which columns can be reordered - is there a UI pattern that would make that clear? Maybe showing the up and down arrows at once greyed out, like this Bootstrap example:

![image](https://github.com/user-attachments/assets/cc18798f-de57-4424-bb02-69a3130cd761)

I think also a hover effect would help.

Lastly, can you make those `<button>` elements? Because otherwise if you use keyboard (e.g. tab through the page) you can't reach the sort buttons and activate them. I'm happy to chat about this at hack night too!

tedschu commented:
@vkoves sure thing - let me look at this early this week and we can also talk it over at the meeting as you mentioned. Thanks!

vkoves commented:
And one thing I forgot to mention, let's make sure there's a text alternative, like using an `aria-label`:

![image](https://github.com/user-attachments/assets/913abd65-58b4-4737-92e6-bab74f1cfa39)

tedschu commented:
@vkoves I have updated the sort elements to be more visible (even when 'inactive'), and added aria labels. If you'd prefer other modifications (ex. the Boostrap version above) happy to do it, but perhaps it'd be good in that case if we could work on this together a bit next week.

vkoves commented on src/components/BuildingsTable.vue:
```diff
@@ -31,6 +31,24 @@ export default class BuildingsTable extends Vue {
+   * based on whether the column is selected and sorted */
+  getAriaLabelForSort(fieldname: string): string {
+    if (fieldname === this.sortedField) {
+      return `Currently sorted by ${this.sortedField} in 
```
Let's tweak this label a bit, as it's confusing - we need to communicate state, but also what will happen. Perhaps "Toggle Sort ${this.sortedField} - Currently Sorting Asc./Desc.". That makes clear if it's descending it'll flip the other way.

![image](https://github.com/user-attachments/assets/82fa4ced-e734-4da9-911e-295c98263eb6)

For the other one, I'd just change it to `Sort By ${this.sortedField}`, because it's shorter, and the column is implied and would be clear to a screen reader (it's in a column heading element). It's a bit harder to explain, so if you're at all interested in learning more about accessibility, I'd encourage downloading and playing with the free [NVDA for Windows](https://www.nvaccess.org/download/) or VoiceOver on Mac ([guide](https://support.apple.com/guide/voiceover/turn-voiceover-on-or-off-vo2682/mac))!

vkoves commented on src/components/BuildingsTable.vue:
```diff
@@ -225,6 +327,21 @@ export default class BuildingsTable extends Vue {
+          border-bottom: none;
+        }
+        .sort.selected {
+          color: black;
```
Can you switch these to use our CSS colors from `colors.scss`?

vkoves commented on src/pages/Search.vue:
```diff
@@ -51,6 +51,16 @@ export default class Search extends Vue {
+  sortedField = 'GHGIntensity';
+
+  /** The direction of the sorted field (column) */
+  sortedDirection = 'desc';
```
If this can only be `'asc'` or 'desc'`, we can type it as such, which adds type safety, by doing:

```ts
sortedDirection: 'asc' | 'desc' = 'desc';
```

That way someone later can't accidentally do `sorteDirection = 'ascending'`.

vkoves commented on src/pages/Search.vue:
```diff
@@ -181,10 +192,56 @@ export default class Search extends Vue {
+
+  /** Called from handleSort, this function sorts
+   * according to sortedField and sortedDirection state values */
+  runSort(buildings: any): void {
```
I think buildings _should_ have the type `Array<IBuildingNode>`, which will then make other stuff easier

vkoves commented on src/components/BuildingsTable.vue:
```diff
@@ -46,21 +64,105 @@ export default class BuildingsTable extends Vue {
+          >
             Fossil Gas Use<br />
             <span class="unit">(kBtu)</span>
+            <button
```
I think you need to add an input to the table for `showSort`, since without it, you can see the sort on pages where it doesn't do anything, like [top emitters](https://deploy-preview-174--radiant-cucurucho-d09bae.netlify.app/top-emitters)!

![image](https://github.com/user-attachments/assets/2c17ae93-d854-4be2-8b99-09a61cd120c4)

vkoves commented on src/components/BuildingsTable.vue:
```diff
@@ -31,6 +31,24 @@ export default class BuildingsTable extends Vue {
   @Prop({ default: false }) showElectricityUse!: boolean;
+
+  /** Props from Search component to store field (column) being sorted and direction */
+  @Prop({ default: 'GHGIntensity' }) sortedField!: string;
```
You can also combine these into one object, if you'd like, like:

```
ts@Prop({ default: defaultSort }) sortParams: { field: string; direction: 'asc' | 'desc' }
```

That's optional and up to you though, but I wanted to mention it - since those two pieces of data are tightly coupled, I think keeping them wrapped would be appropriate.

tedschu replied:
You're right on this. If you don't mind, I'd like to quickly discuss live next time we're meeting to review this.

vkoves commented on src/components/BuildingsTable.vue:
```diff
@@ -31,6 +31,24 @@ export default class BuildingsTable extends Vue {
+  @Prop({ default: 'desc' }) sortedDirection!: string;
+
+  // declares the property emit for TS
+  $emit: any;
```
Interesting, I'm a little confused what's wrong here, but I'll need to dig into that more later.

Could you add a todo around this? (e.g. `// TODO: Figure out why Vue `$emit` isn't recognized by Typescript`). You shouldn't have to add this hack, so it's probably something I misconfigured 😅

vkoves reviewed (changes requested):
Alright, took some time to review and I have a few minor comments, but this looks great overall!

vkoves commented:
I found some nice UI from Material, going to try and get a bit of this in here: https://v5.material.angular.io/components/table/overview#sorting

vkoves reviewed (approved):
Awesome work on this, thanks Ted! I made some minor cosmetic tweaks to add a little hover effect and make sure the `cursor: pointer` is used on the entire column header, so it's clear you can click the whole thing
